File: mgyoutube-multitech/api-dotnet/repos/CouchSearchesDataRepoImpl.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using api_dotnet.couchutils;
using Couchbase;
using Couchbase.Authentication;
using Couchbase.Configuration.Client;
using Couchbase.Core;
using Couchbase.Core.Buckets;
using Couchbase.Management;
using Couchbase.N1QL;

namespace api_dotnet.repos
{
    public class CouchSearchesDataRepoImpl : SearchesDataRepo
    {
        private readonly Cluster cluster;
        private readonly string username;
        private readonly string password;

        public const string SEARCHES_BUCKET_NAME = "searches";
        public const string SEARCHES_PARENT_INDEX_NAME = "searches-parentuserid-idx";
        public const string SEARCHES_BUCKET_PARENT_FIELDNAME = "parentUserId";
        public const string SEARCHES_BUCKET_SEARCH_PHRASE_FIELDNAME = "phrase";

        public const string SEARCHES_FIND_PHRASE_BY_PARENT = "SELECT phrase FROM `searches` WHERE parentUserId = $1";
        public const string SEARCHES_DELETE_PHRASE_AND_PARENT = "DELETE FROM `searches` WHERE parentUserId = $1 AND phrase = $2";

        public CouchSearchesDataRepoImpl(string connectionString, string username, string password)
        {
            this.username = username;
            this.password = password;
            this.cluster = new Cluster(new ClientConfiguration
            {
                Servers = new List<Uri> { new Uri(connectionString) }
            });

            IAuthenticator authenticator = new PasswordAuthenticator(username, password);
            cluster.Authenticate(authenticator);
        }
        public void RepositoryStartup()
        {
            Console.WriteLine("CouchSearchesDataRepoImpl.RepositoryStartup");

            IClusterManager clusterManager = this.cluster.CreateManager(this.username, this.password);

            if (!CouchUtils.DoesBucketExist(clusterManager, SEARCHES_BUCKET_NAME))
            {
                IResult createBucketResults = clusterManager.CreateBucket(SEARCHES_BUCKET_NAME,
                    bucketType: BucketTypeEnum.Couchbase,
                    ramQuota: 128, // megabytes
                    replicaNumber: ReplicaNumber.One,
                    indexReplicas: true,
                    flushEnabled: true);

                if (!createBucketResults.Success)
                {
                    throw new Exception(createBucketResults.Message);
                }
                IBucket searchesBucket = cluster.OpenBucket(SEARCHES_BUCKET_NAME);
                IBucketManager searchesBucketManager = searchesBucket.CreateManager(this.username, this.password);
                IResult createPrimaryIndexResult = searchesBucketManager.CreateN1qlPrimaryIndex(false);
                if (!createPrimaryIndexResult.Success)
                {
                    throw new Exception(createPrimaryIndexResult.Message);
                }
                string[] fields = { SEARCHES_BUCKET_PARENT_FIELDNAME };
                IResult createParentsIndexResult = searchesBucketManager.CreateN1qlIndex(SEARCHES_PARENT_INDEX_NAME, false, fields);
                if (!createParentsIndexResult.Success)
                {
                    throw new Exception(createParentsIndexResult.Message);
                }
            }
            else
            {
                IBucket searchesBucket = cluster.OpenBucket(SEARCHES_BUCKET_NAME);
                IBucketManager searchesBucketManager = searchesBucket.CreateManager(this.username, this.password);
                Console.WriteLine("flushing " + SEARCHES_BUCKET_NAME);
                IResult flushResults = searchesBucketManager.Flush();
                if (!flushResults.Success)
                {
                    throw new Exception(flushResults.Message);
                }
            }
        }

        public async Task AddSearchToParentUser(string parentUserId, string searchPhrase)
        {
            string id = Guid.NewGuid().ToString();
            Document<dynamic> document = new Document<dynamic> { Id = id, Content = new { parentUserId = parentUserId, phrase = searchPhrase } };

            IBucket searchesBucket = cluster.OpenBucket(SEARCHES_BUCKET_NAME);
            IDocumentResult result = await searchesBucket.InsertAsync(document);

            Console.WriteLine("Couchdb inserted parentUserId " + parentUserId + " and searchPhrase " + searchPhrase
                    + ", docId: " + id + " result.Message: " + result.Message);
        }

        public async Task<List<string>> GetSearchesForParentUser(string parentUserId)
        {
            Console.WriteLine("Couchdb searching " + SEARCHES_BUCKET_NAME + " for parentUserId " + parentUserId);

            List<string> searches = new List<string>();

            QueryRequest query = new QueryRequest(SEARCHES_FIND_PHRASE_BY_PARENT);
            query.ScanConsistency(ScanConsistency.RequestPlus);
            query.AddPositionalParameter(parentUserId);

            IBucket searchesBucket = cluster.OpenBucket(SEARCHES_BUCKET_NAME);
            using (IQueryResult<dynamic> results = await searchesBucket.QueryAsync<dynamic>(query))
            {
                if (!results.Success)
                {
                    Console.WriteLine("Couchdb failed to search for " + parentUserId + " Message: " + results.Message + " Status:"
                            + results.Status);
                    results.Errors.ForEach(e =>
                    {
                        Console.WriteLine("Error Code: " + e.Code + " Severity: " + e.Severity + " Name: " + e.Name + " Message: " + e.Message);
                    });

                    Console.WriteLine("results.Exception: " + results.Exception);
                    if (results.Exception != null)
                    {
                        throw results.Exception;
                    }
                    throw new Exception("Unable to get any searches for parentUserId " + parentUserId);
                }

                foreach (dynamic row in results)
                {
                    string phrase = row.phrase;
                    searches.Add(phrase);
                }
            }

            Console.WriteLine("Couchdb searching " + SEARCHES_BUCKET_NAME + " for parentUserId " + parentUserId
                    + " returning " + searches.Count + " phrases");
            return searches;
        }

        public async Task RemoveSearchFromParentUser(string parentUserId, string searchPhrase)
        {
            QueryRequest query = new QueryRequest(SEARCHES_DELETE_PHRASE_AND_PARENT);
            query.ScanConsistency(ScanConsistency.RequestPlus);
            query.AddPositionalParameter(parentUserId);
            query.AddPositionalParameter(searchPhrase);
            IBucket searchesBucket = cluster.OpenBucket(SEARCHES_BUCKET_NAME);

            using (IQueryResult<dynamic> results = await searchesBucket.QueryAsync<dynamic>(query))
            {
                Console.WriteLine("Couchdb deleted parentUserId " + parentUserId + " and searchPhrase " + searchPhrase + ": "
                        + results.Status);
            }
        }
    }
}

File: mgyoutube-multitech/api-python/repos/CouchSearchesDataRepoImpl.py
import uuid
from repos.SearchesDataRepo import SearchesDataRepo

from couchbase.cluster import Cluster
from couchbase.cluster import PasswordAuthenticator
from couchbase.n1ql import N1QLQuery, CONSISTENCY_REQUEST
from couchbase.admin import Admin
from couchutils.couchutils import doesBucketExist, createBasicBucket


class CouchSearchesDataRepoImpl(SearchesDataRepo):
    SEARCHES_BUCKET_NAME = "searches"
    SEARCHES_PARENT_INDEX_NAME = "searches-parentuserid-idx"
    SEARCHES_BUCKET_PARENT_FIELDNAME = "parentUserId"
    SEARCHES_BUCKET_SEARCH_PHRASE_FIELDNAME = "phrase"

    SEARCHES_FIND_PHRASE_BY_PARENT = "SELECT phrase FROM `searches` WHERE parentUserId = $1"
    SEARCHES_DELETE_PHRASE_AND_PARENT = "DELETE FROM `searches` WHERE parentUserId = $1 AND phrase = $2"

    def __init__(self, connectionString: str, username: str, password: str):
        # self.username = username
        # self.password = password
        # self.connectionString = connectionString
        self.cluster = Cluster(connectionString)
        authenticator = PasswordAuthenticator(username, password)
        self.cluster.authenticate(authenticator)

    def repositoryStartup(self):
        if (not doesBucketExist(CouchSearchesDataRepoImpl.SEARCHES_BUCKET_NAME, self.cluster)):
            searchesBucket = createBasicBucket(
                CouchSearchesDataRepoImpl.SEARCHES_BUCKET_NAME, self.cluster)
            # self.cluster.cluster_manager().bucket_create(
            #     CouchSearchesDataRepoImpl.SEARCHES_BUCKET_NAME, 'couchbase', "", 0, 128, True)
            # self.cluster.cluster_manager().wait_ready(
            #     CouchSearchesDataRepoImpl.SEARCHES_BUCKET_NAME, 8)
            # searchesBucket = self.cluster.open_bucket(
            #     CouchSearchesDataRepoImpl.SEARCHES_BUCKET_NAME)
            bucketManager = searchesBucket.bucket_manager()
            # bucketManager.create_n1ql_primary_index()
            bucketManager.create_n1ql_index(
                CouchSearchesDataRepoImpl.SEARCHES_PARENT_INDEX_NAME, defer=False, ignore_exists=True, fields=[CouchSearchesDataRepoImpl.SEARCHES_BUCKET_PARENT_FIELDNAME])
        else:
            searchesBucket = self.cluster.open_bucket(
                CouchSearchesDataRepoImpl.SEARCHES_BUCKET_NAME)
            searchesBucket.flush()

    def getSearchesForParentUser(self, parentUserId: str) -> list:
        searches = list()
        query = N1QLQuery(
            CouchSearchesDataRepoImpl.SEARCHES_FIND_PHRASE_BY_PARENT, parentUserId)
        query.consistency = CONSISTENCY_REQUEST
        searchesBucket = self.cluster.open_bucket(
            CouchSearchesDataRepoImpl.SEARCHES_BUCKET_NAME)
        for row in searchesBucket.n1ql_query(query):
            phrase: str = row[CouchSearchesDataRepoImpl.SEARCHES_BUCKET_SEARCH_PHRASE_FIELDNAME]
            searches.append(phrase)
        return searches

    def addSearchToParentUser(self, parentUserId: str, searchPhrase: str) -> None:
        id = str(uuid.uuid4())
        doc = dict()
        doc[CouchSearchesDataRepoImpl.SEARCHES_BUCKET_PARENT_FIELDNAME] = parentUserId
        doc[CouchSearchesDataRepoImpl.SEARCHES_BUCKET_SEARCH_PHRASE_FIELDNAME] = searchPhrase

        searchesBucket = self.cluster.open_bucket(
            CouchSearchesDataRepoImpl.SEARCHES_BUCKET_NAME)
        insertRv = searchesBucket.insert(id, doc)
        if (not insertRv.success):
            print("CouchSearchesDataRepoImpl.addSearchToParentUser failed", insertRv.rc)

    def removeSearchFromParentUser(self, parentUserId: str, searchPhrase: str):
        query = N1QLQuery(CouchSearchesDataRepoImpl.SEARCHES_DELETE_PHRASE_AND_PARENT,
                          parentUserId,  searchPhrase)
        searchesBucket = self.cluster.open_bucket(
            CouchSearchesDataRepoImpl.SEARCHES_BUCKET_NAME)
        searchesBucket.n1ql_query(query).execute()
